per_channel fake quant fp16 and fp64 support

HDCharles opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56894 per_channel fake quant fp16 and fp64 support**

Summary: used the dispatch type macro to add support for fp16 and 64 tensors. 

Test Plan:
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_half_precision_numerics
python test/test_quantization.py TestFakeQuantize
python test/test_quantization.py TestFakeQuantize.test_backward_per_channel_cachemask_cpu
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_cachemask_cpu

tested in fbcode on-demand gpu:
buck build @mode/dev-nosan //caffe2:torch
buck test @mode/opt //caffe2/test:quantization --test-selectors test_workflow_module#TestFakeQuantize
result: https://www.internalfb.com/intern/testinfra/testconsole/testrun/2533274846016159/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28002955](https://our.internmc.facebook.com/intern/diff/D28002955)

HDCharles committed:
per_channel fake quant fp16 and fp64 support

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vadimkantorov commented on test/quantization/test_workflow_module.py:
```diff
@@ -1384,7 +1419,8 @@ def _test_backward_per_channel_cachemask_impl(self, device):
                 X, scale, zero_point, axis, quant_min, quant_max)
-            dout = torch.rand(X.shape, dtype=torch.float).to(device)
+            # dout = torch.rand(X.shape, dtype=torch.float).to(device)
+            dout = torch.rand(X.shape, dtype=float_type).to(device)
```
should this be `torch.rand_like(X, dtype=float_type, device=device)`?

HDCharles replied:
I can make this change although I'm wondering if there is a functional difference? Is the rand_like considered a best practice?

vadimkantorov replied:
i think mostly matter of conciseness / style + maybe some savings on creating a random tensor on the target device directly (actually not sure if it's supported on all devices?)

vkuzo commented on aten/src/ATen/native/quantized/fake_quant_per_channel_affine.cpp:
```diff
@@ -48,9 +48,9 @@ std::tuple<Tensor, Tensor> fake_quantize_per_channel_affine_cachemask(
-  TORCH_CHECK(self.scalar_type() == ScalarType::Float);
-  TORCH_CHECK(scale.scalar_type() == ScalarType::Float,
-              "Scale must be Float, found ", scale.scalar_type());
+  // TORCH_CHECK(self.scalar_type() == ScalarType::Float);
```
nit: we can remove these

HDCharles replied:
will do

vkuzo commented on test/quantization/test_workflow_module.py:
```diff
@@ -117,27 +117,29 @@ def _permute_to_axis_zero(X, axis):
 def _fake_quantize_per_channel_affine_reference(X, per_channel_scale, per_channel_zero_point, axis, quant_min, quant_max):
-    X, permute_axis_list = _permute_to_axis_zero(X, axis)
+    dtype = X.dtype
+    X, permute_axis_list = _permute_to_axis_zero(X.to(torch.float32), axis)
```
do I understand correctly that from L121 to L128, `X` has dtype of `torch.float`?  If yes, should that be changed?  I would have expected the dtype to be consistent with dtype under test throughout the test case.

HDCharles replied:
the issue is the reference function doesn't support half precision floats. So I have to convert to float32. 

This is also what we did with per_tensor: https://github.com/pytorch/pytorch/blob/a90a3acbeec7deebecf7baf53c8f0fdf7b37e5ae/test/quantization/test_workflow_module.py#L57

is there a different way to implement the reference function that I should be using?

vkuzo replied:
which specific ops do not support fp32?  Can we just leave those ops in fp32, and do the rest in fp16?

HDCharles replied:
rounding is the main issue, I don't see how it changes anything though, if I added .to(torch.float) within the calculation for res, its funcitonally the same, right?

vkuzo replied:
IMO it's not worth it to implement fp16 support for torch.round just to make this test pass, so this reference function isn't really the correct answer.

I do think it's worth minimizing the differences between what this function is doing comparing to what it could do if all ops supported fp16.  It could be something like

```
# old
 res[i] = (torch.clamp(torch.round(X[i] * (1.0 / per_channel_scale[i]) + per_channel_zero_point[i]), quant_min, quant_max) - per_channel_zero_point[i]) * per_channel_scale[i]

# new
 scaled_correct_dtype = X[i] * (1.0 / per_channel_scale[i]) + per_channel_zero_point[i]
 scaled_fp32 = scaled_correct_dtype.float()
 rounded = torch.round(scaled_fp32)
 rounded_correct_dtype = rounded.to(correct_dtype)
 res[i] = (torch.clamp(rounded_correct_dtype, quant_min, quant_max) - per_channel_zero_point[i]) * per_channel_scale[i]
```

HDCharles replied:
note: clamp is also not supported for fp16

vkuzo replied:
ok, then how about

```
# torch.clamp is not supported in fp16
def _clamp_in_float(x, min_val, max_val):
  old_dtype = x.dtype
  x = x.to(torch.float)
  x = torch.clamp(x, min_val, max_val)
  x = x.to(old_dtype)
  return x

# torch.round is not supported in fp16
def _round_in_float(x, min_val, max_val):
  old_dtype = x.dtype
  x = x.to(torch.float)
  x = torch.round(x)
  x = x.to(old_dtype)
  return x

res[i] = (_clamp_in_float(_round_in_float(X[i] * (1.0 / per_channel_scale[i]) +
                  per_channel_zero_point[i]), quant_min, quant_max) - per_channel_zero_point[i]) * per_channel_scale[i]
```

vkuzo replied:
@HDCharles found that this is due to the scale being in fp32.  The current behavior is expected then.  Would be great to add a comment to the code for future readers.

HDCharles committed:
Update on "per_channel fake quant fp16 and fp64 support"


Summary: used the dispatch type macro to add support for fp16 and 64 tensors. 

Test Plan:
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_half_precision_numerics
python test/test_quantization.py TestFakeQuantize
python test/test_quantization.py TestFakeQuantize.test_backward_per_channel_cachemask_cpu
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_cachemask_cpu

tested in fbcode on-demand gpu:
buck build @mode/dev-nosan //caffe2:torch
buck test @mode/opt //caffe2/test:quantization --test-selectors test_workflow_module#TestFakeQuantize
result: https://www.internalfb.com/intern/testinfra/testconsole/testrun/2533274846016159/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28002955](https://our.internmc.facebook.com/intern/diff/D28002955)

[ghstack-poisoned]

HDCharles committed:
Update on "per_channel fake quant fp16 and fp64 support"


Summary: used the dispatch type macro to add support for fp16 and 64 tensors. 

Test Plan:
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_half_precision_numerics
python test/test_quantization.py TestFakeQuantize
python test/test_quantization.py TestFakeQuantize.test_backward_per_channel_cachemask_cpu
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_cachemask_cpu

tested in fbcode on-demand gpu:
buck build @mode/dev-nosan //caffe2:torch
buck test @mode/opt //caffe2/test:quantization --test-selectors test_workflow_module#TestFakeQuantize
result: https://www.internalfb.com/intern/testinfra/testconsole/testrun/2533274846016159/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28002955](https://our.internmc.facebook.com/intern/diff/D28002955)

[ghstack-poisoned]

HDCharles committed:
Update on "per_channel fake quant fp16 and fp64 support"


Summary: used the dispatch type macro to add support for fp16 and 64 tensors. 

Test Plan:
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_half_precision_numerics
python test/test_quantization.py TestFakeQuantize
python test/test_quantization.py TestFakeQuantize.test_backward_per_channel_cachemask_cpu
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_cachemask_cpu

tested in fbcode on-demand gpu:
buck build @mode/dev-nosan //caffe2:torch
buck test @mode/opt //caffe2/test:quantization --test-selectors test_workflow_module#TestFakeQuantize
result: https://www.internalfb.com/intern/testinfra/testconsole/testrun/2533274846016159/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28002955](https://our.internmc.facebook.com/intern/diff/D28002955)

[ghstack-poisoned]

HDCharles committed:
Update on "per_channel fake quant fp16 and fp64 support"


Summary: used the dispatch type macro to add support for fp16 and 64 tensors. 

Test Plan:
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_half_precision_numerics
python test/test_quantization.py TestFakeQuantize
python test/test_quantization.py TestFakeQuantize.test_backward_per_channel_cachemask_cpu
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_cachemask_cpu

tested in fbcode on-demand gpu:
buck build @mode/dev-nosan //caffe2:torch
buck test @mode/opt //caffe2/test:quantization --test-selectors test_workflow_module#TestFakeQuantize
result: https://www.internalfb.com/intern/testinfra/testconsole/testrun/2533274846016159/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28002955](https://our.internmc.facebook.com/intern/diff/D28002955)

[ghstack-poisoned]

HDCharles committed:
Update on "per_channel fake quant fp16 and fp64 support"


Summary: used the dispatch type macro to add support for fp16 and 64 tensors. 

Test Plan:
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_half_precision_numerics
python test/test_quantization.py TestFakeQuantize
python test/test_quantization.py TestFakeQuantize.test_backward_per_channel_cachemask_cpu
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_cachemask_cpu

tested in fbcode on-demand gpu:
buck build @mode/dev-nosan //caffe2:torch
buck test @mode/opt //caffe2/test:quantization --test-selectors test_workflow_module#TestFakeQuantize
result: https://www.internalfb.com/intern/testinfra/testconsole/testrun/2533274846016159/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28002955](https://our.internmc.facebook.com/intern/diff/D28002955)

[ghstack-poisoned]

HDCharles committed:
Update on "per_channel fake quant fp16 and fp64 support"


Summary: used the dispatch type macro to add support for fp16 and 64 tensors. 

Test Plan:
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_half_precision_numerics
python test/test_quantization.py TestFakeQuantize
python test/test_quantization.py TestFakeQuantize.test_backward_per_channel_cachemask_cpu
python test/test_quantization.py TestFakeQuantize.test_forward_per_channel_cachemask_cpu

tested in fbcode on-demand gpu:
buck build @mode/dev-nosan //caffe2:torch
buck test @mode/opt //caffe2/test:quantization --test-selectors test_workflow_module#TestFakeQuantize
result: https://www.internalfb.com/intern/testinfra/testconsole/testrun/2533274846016159/

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28002955](https://our.internmc.facebook.com/intern/diff/D28002955)

[ghstack-poisoned]
